cloud/docker_facts.py

```diff
@@ -0,0 +1,246 @@
+        self.container_id = module.params.get('id')
+        self.name = module.params.get('name')
+        self.images = module.params.get('images')
+        self.all_containers = not (self.name or self.container_id)
+
+        try:
+            self.client = docker.Client(base_url=docker_url.geturl())
+        except DockerAPIError, e:
+            module.fail_json(msg="Docker API error: " + e.explanation)
+        except RequestException, e:
+            module.fail_json(msg="Failed to instantiate docker client. This "
+                                 "could mean that docker isn't running.: %s" %
+                                 repr(e))
+
+    def docker_facts_slugify(self, value):
+        return 'docker_%s' % (re.sub('[^\w-]', '_', value).lower().lstrip('_'))
+
+    def get_inspect(self, id, type='container'):
+        facts = dict()
+        if type == 'images':
+            inspect = self.client.inspect_image(id)
+        else:
+            inspect = self.client.inspect_container(id)
+
+        for key in inspect:
+            fact_key = self.docker_facts_slugify(key)
+            facts[fact_key] = inspect.get(key)
+
+        facts['docker_short_id'] = id[:13]
+        return facts
+
+    def entity_conform(self, entity):
+        new_entity = dict()
+
+        for key in entity:
+            new_key = self.docker_facts_slugify(key)
+            new_entity[new_key] = entity.get(key)
+
+        return new_entity
+
+    def search(self, entities, type='containers'):
+        search_list = []
+
+        if type == 'images':
+            id = self.images
+            name = None
+        else:
+            id = self.container_id
+            name = self.name
+
+        for entity in entities:
+            entity_name = ''
+            # by name, get by name
+            if name:
+                # names have a leading '/'
+                if not name.startswith('/'):
+                    entity_name = '/' + name
+                if (entity_name in entity['Names']):
+                    search_list.append(entity)
+            # by name, get by name
+            elif id:
+                # container_id could be any length, but internally
+                # with docker is 64 chars
+                if id in entity['Id']:
+                    search_list.append(entity)
+
+        # an ID should never yield more than one, but in case
+        if len(entity) > 1 and id and type == 'containers':
+            module.fail_json(msg='Multiple items found for container id '
+                                 '%s' % id)
+
+        return search_list
+
+    def get_facts(self, entities, type='containers'):
+        facts = dict()
+        entity_dict = dict()
+        for entity in entities:
+            id = entity.get('Id')
+            short_id = id[:13]
+            name = short_id
+
+            if type == 'containers':
+                try:
+                    name = entity.get('Names', list()).pop(0).lstrip('/')
+                except IndexError:
```

Comment:
Just add `AttributeError` so that if the resulting `list()` object is defaulted to, then we catch `.pop(0)` correctly and then set `name = short_name`.